libs/config/convert/to_typed.go

```diff
@@ -0,0 +1,223 @@
+package convert
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"github.com/databricks/cli/libs/config"
+)
+
+func ToTyped(dst any, src config.Value) error {
+	dstv := reflect.ValueOf(dst)
+
+	// Dereference pointer if necessary
+	for dstv.Kind() == reflect.Pointer {
+		if dstv.IsNil() {
+			dstv.Set(reflect.New(dstv.Type().Elem()))
+		}
+		dstv = dstv.Elem()
+	}
+
+	// Verify that vv is settable.
+	if !dstv.CanSet() {
+		panic("cannot set destination value")
+	}
+
+	switch dstv.Kind() {
+	case reflect.Struct:
+		return toTypedStruct(dstv, src)
+	case reflect.Map:
+		return toTypedMap(dstv, src)
+	case reflect.Slice:
+		return toTypedSlice(dstv, src)
+	case reflect.String:
+		return toTypedString(dstv, src)
+	case reflect.Bool:
+		return toTypedBool(dstv, src)
+	case reflect.Int, reflect.Int32, reflect.Int64:
+		return toTypedInt(dstv, src)
+	case reflect.Float32, reflect.Float64:
+		return toTypedFloat(dstv, src)
+	}
+
+	return fmt.Errorf("unsupported type: %s", dstv.Kind())
+}
+
+func toTypedStruct(dst reflect.Value, src config.Value) error {
+	switch src.Kind() {
+	case config.KindMap:
+		info := getStructInfo(dst.Type())
+		for k, v := range src.MustMap() {
+			index, ok := info.Fields[k]
+			if !ok {
+				// unknown field
+				continue
```

Comment:
shouldn't we return an error if the field is unknown?

Reply:
It should be a warning (over time). It is allowed to have anything in YAML, define it as an anchor, and then reuse it across the rest of the document. This should not be an error, but a warning at best.

The code here only errors on things that are hard errors. Collecting all warnings is a separate goal (and something we should add in the very near term), IMO.

Reply:
Updated the comment to point to this PR and context.